File: proxy-poc2/src/poc/handler.ts
import { IncomingMessage, ServerResponse } from "http";
import { Proxy } from "./proxy";
import { parseRequest } from "./parser";

export function customListener(legacy: string, release: string) {
  const source = new Proxy(new URL(legacy));
  const target = new Proxy(new URL(release));

  return async (req: IncomingMessage, res: ServerResponse) => {
    const { body, group, status } = await parseRequest(req);
    const [leader, shadow] = status === "COMPLETE" ? [target, source] : [source, target];

    leader.modifyResponse = (res: ServerResponse) => {
      res.setHeader("X-goproxy", `group=${group};status=${status}`);
    };
    shadow.modifyResponse = undefined;

    await Promise.all([
      // leader.serveProxy(req, res, body),
      shadow.serveProxy(clone(req), res, body),
    ]);
  };

  function clone<T>(dolly: T): T {
    return { ...dolly } as T;
  }
}

File: proxy-poc/poc/handler.go
package poc

import (
	"context"
	"fmt"
	"net/http"
)

func CustomHandler(from, to string) http.HandlerFunc {
	source := NewProxy(from)
	target := NewProxy(to)

	return func(w http.ResponseWriter, r *http.Request) {
		body, group, status, _ := parseRequest(r)
		leader, shadow := source, target
		if status == "COMPLETE" {
			leader, shadow = target, source
		}

		leader.ModifyResponse = func(res *http.Response) error {
			res.Header.Set("X-goproxy", fmt.Sprintf("group=%s;status=%s", group, status))
			return nil
		}
		shadow.ModifyResponse = nil

		leader.ServeProxy(w, r, body)
		shadow.ServeProxy(dummyResponseWriter{}, r.Clone(context.TODO()), body)
	}
}
